### DISCUSSION TITLE
Support for Cloudflare R2 (S3 compatible object storage)

### DISCUSSION THREAD

**Discussion by Chinoman10** [langfuse/langfuse#4788] [Ideas]
### Describe the feature or potential improvement

The current V3 architecture uses AWS's S3 object storage for storing "multi-modal data and raw tracing events".

I'd just like to suggest having [Cloudflare's R2](https://www.cloudflare.com/en-gb/developer-platform/products/r2/) (_S3-compatible_) as a cheaper alternative (_no egress charges, among other benefits_).

**Comment by marcklingen** [comment]
Thanks for raising this. You should be able to use Cloudflare R2 as they are S3 compatible. In fact, we have used Cloudflare R2 ourselves in the past for Langfuse Cloud before consolidating all infrastructure on AWS.

More details here: https://langfuse.com/self-hosting/infrastructure/blobstorage

If you think that additional notes on R2 would be helpful here, any contributions to the documentation is highly appreciated.

**Reply by Chinoman10** [reply]
Oh cool, didn't notice the docs already reflected it :) (_we're not self-hosting Langfuse ourselves, so I never looked into those docs_ 😅 ).

**Reply by marcklingen** [reply]
Nice, if you do not self-host, you do not need to worry about S3/R2 :)

**Reply by Chinoman10** [reply]
Lol, yes I know that of course :) but wanted to leave the suggestion anyway as I very often hear complaints about S3's egregious pricing and when people switch over to R2 they often don't look back. So this was both a suggestion for the Langfuse team to keep costs down, but also for the people that self-host it ^^ it wasn't something for me/us specifically :)

**Reply by marcklingen** [reply]
makes sense, thanks for sharing! we use R2 for many static assets on langfuse.com and it is great

**Reply by Steffen911** [reply]
Also, I'll cross-post my response from Discord as it's also applicable to this discussion:
> Yes, that was a factor that we considered in our decision. As far as we could tell there was about a 10-20% discount on API operations with R2. This discount was not enough to offset the egress costs that we anticipated from the AWS side if the traffic leaves our VPC to be written into R2. In addition, we can use IAM roles on our services to avoid long-living credentials by staying on S3. So for our cloud environment, S3 seems like the better choice.
>
> For self-hosters both options are perfectly valid and supported. While the env variable naming is S3 first, we support multiple storage backends as discussed in https://langfuse.com/self-hosting/infrastructure/blobstorage.

https://discord.com/channels/1111061815649124414/1111063198989307964/1319292484915826689
